### ISSUE TITLE
pmca-console doesn't (fully?) detect a Sony FDR-X1000V

### ISSUE THREAD

**Issue by enleth** [ma1co/Sony-PMCA-RE#25]
tl;dr it detects a Sony device based on the vendor ID and starts querying the mass storage interface and but doesn't recognize it as a device it can use. The camera resets its USB interface after querying.

FYI I'm running pmca on Linux, which seems to not be among your target platforms, but it doesn't appear to be a source of my problem, as far as I can tell.

My goal here is to initiate a software update with a custom-crafted updater image and no firmware payload to get some sort of a shell on the camera and ideally avoid bricking it in the process, and your tools just seemed to be the ideal means of doing that, with fwtool.py and this. I've seen #20 which probably applies here as well but I don't care about installing regular apps.

I should be able to figure it out mostly by myself and contribute some code in the end, but I need some guidance from you to get started efficiently - I've done my share of embedded development and reverse engineering and I know python pretty well, but that's the first time I'd be playing with custom USB mass storage commands in general and Sony cameras in particular.

The FDR-X1000V registers two separate endpoints on a single mass storage interface, one of those is the microSD card, the other is a small read-only vfat filesystem that contains the license files and WiFi connection details. To the best of my knowledge, it does not support MTP. Probing it with pmca_console info does _something_, as the camera does a USB disconnect or reset (seen both, no idea why), but that's pretty much it.

This is what I do and what happens:

```
enleth@panzerbook ~/sony/Sony-PMCA-RE $ dmesg
...
[20611.202184] usb 2-1.2: new high-speed USB device number 8 using xhci_hcd
[20611.332991] usb 2-1.2: New USB device found, idVendor=054c, idProduct=07f5
[20611.332994] usb 2-1.2: New USB device strings: Mfr=1, Product=2, SerialNumber=3
[20611.332995] usb 2-1.2: Product: FDR-X1000V
[20611.332997] usb 2-1.2: Manufacturer: Sony
[20611.332998] usb 2-1.2: SerialNumber: 6C092035EFAB
[20611.334050] usb-storage 2-1.2:1.0: USB Mass Storage device detected
[20611.362166] scsi host4: usb-storage 2-1.2:1.0
[20612.412823] scsi 4:0:0:0: Direct-Access     Sony     Camcorder        1.00 PQ: 0 ANSI: 0
[20612.412960] sd 4:0:0:0: Attached scsi generic sg2 type 0
[20612.617673] scsi 4:0:0:1: Direct-Access     Sony     Camcorder        1.00 PQ: 0 ANSI: 0
[20612.617896] sd 4:0:0:1: Attached scsi generic sg3 type 0
[20612.619500] sd 4:0:0:0: [sdc] 249737216 512-byte logical blocks: (128 GB/119 GiB)
[20612.620141] sd 4:0:0:0: [sdc] Write Protect is on
[20612.620144] sd 4:0:0:0: [sdc] Mode Sense: 03 00 80 00
[20612.620585] sd 4:0:0:0: [sdc] No Caching mode page found
[20612.620587] sd 4:0:0:0: [sdc] Assuming drive cache: write through
[20612.621435] sd 4:0:0:1: [sdd] 4096 512-byte logical blocks: (2.10 MB/2.00 MiB)
[20612.621834] sd 4:0:0:1: [sdd] Write Protect is on
[20612.621837] sd 4:0:0:1: [sdd] Mode Sense: 03 00 80 00
[20612.622231] sd 4:0:0:1: [sdd] No Caching mode page found
[20612.622234] sd 4:0:0:1: [sdd] Assuming drive cache: write through
[20612.639899]  sdd: sdd1
[20612.646106]  sdc: sdc1
[20612.647983] sd 4:0:0:1: [sdd] Attached SCSI removable disk
[20612.720252] sd 4:0:0:0: [sdc] Attached SCSI removable disk
enleth@panzerbook ~/sony/Sony-PMCA-RE $ . .env/bin/activate
(.env) enleth@panzerbook ~/sony/Sony-PMCA-RE $ python pmca-console.py info
No native drivers available
Using drivers libusb-MSC, libusb-MTP
Looking for Sony devices

Querying mass storage device

No devices found. Ensure your camera is connected.
(.env) enleth@panzerbook ~/sony/Sony-PMCA-RE $ dmesg
...
[20666.450018] usb 2-1.2: reset high-speed USB device number 8 using xhci_hcd
```

USB device info:

```(.env) enleth@panzerbook ~/sony/Sony-PMCA-RE $ lsusb -d 054c: -v

Bus 002 Device 009: ID 054c:07f5 Sony Corp. 
Device Descriptor:
  bLength                18
  bDescriptorType         1
  bcdUSB               2.00
  bDeviceClass            0 
  bDeviceSubClass         0 
  bDeviceProtocol         0 
  bMaxPacketSize0        64
  idVendor           0x054c Sony Corp.
  idProduct          0x07f5 
  bcdDevice            1.00
  iManufacturer           1 Sony
  iProduct                2 FDR-X1000V
  iSerial                 3 6C092035EFAB
  bNumConfigurations      1
  Configuration Descriptor:
    bLength                 9
    bDescriptorType         2
    wTotalLength           32
    bNumInterfaces          1
    bConfigurationValue     1
    iConfiguration          0 
    bmAttributes         0x80
      (Bus Powered)
    MaxPower              500mA
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        0
      bAlternateSetting       0
      bNumEndpoints           2
      bInterfaceClass         8 Mass Storage
      bInterfaceSubClass      6 SCSI
      bInterfaceProtocol     80 Bulk-Only
      iInterface              0 
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x81  EP 1 IN
        bmAttributes            2
          Transfer Type            Bulk
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0200  1x 512 bytes
        bInterval               0
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x02  EP 2 OUT
        bmAttributes            2
          Transfer Type            Bulk
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0200  1x 512 bytes
        bInterval               0
Device Qualifier (for other device speed):
  bLength                10
  bDescriptorType         6
  bcdUSB               2.00
  bDeviceClass            0 
  bDeviceSubClass         0 
  bDeviceProtocol         0 
  bMaxPacketSize0        64
  bNumConfigurations      1
can't get debug descriptor: Resource temporarily unavailable
Device Status:     0x0000
  (Bus Powered)
```

I can try to run the official firmware update tool in a VM using VirtualBox' USB forwarding and dump the packets exchanged on the host if you suspect this might be helpful.

I could also set up a Linux box with the camera connected to it by USB and give you a shell access if you wanted to poke it yourself.

So - what are the next steps I should take to possibly get it to work and what sort of information should I provide to you in addition to the above?

**Comment by ma1co** [comment]
We detect cameras by a simple string comparison. Try replacing the return statement in `isSonyMscCamera()` in _pmca/usb/sony.py_ by a `return True`.

To run code on the camera, here's a quick description what you can do:
* Run `fwtool.py unpack -f Update.exe -o somedir` on an official firmware update. This creates _somedir/config.yaml_ which contains the camera's model number
* Increase the firmware version in _config.yaml_ (otherwise you won't be able to install it)
* Run `make` in [this](https://github.com/ma1co/OpenMemories-Platform/tree/master/examples/updater) folder. This will compile _main.cpp_ into _build/libupdaterbody_gen3.so_
* Run `fwtool.py pack -c config.yaml -b libupdaterbody_gen3.so -o somedir`. This will pack your updater body into _somedir/firmware_packed.dat_
* Run `pmca-console firmware -f firmware_packed.dat` to "install" the firmware

If everything works out, this should run `UpdaterBodyImpl::Execute()` (which does nothing) and then reboot the camera. The internal firmware of the camera won't be changed (and the firmware number stays the same).

Let me know if you need any help.

**Comment by enleth** [comment]
Thanks, I'm getting somewhere now, although the USB interface appears somewhat unreliable and probably needs some more changes to get it to work with this camera. Also, how do we know it's a gen3 device? I've got the official firmware update unpacked, if there's an indicator of platform version in the sonames or something like that, I can take a look.

So far I've noticed these behaviors:
1. Camera connected and turned on, displays "USB", running `pmca-console firmware -f firmware_packed.dat`, detected the camera right away but hangs on "Initializing firmware update", I ^C'd it after 15 minutes. Camera still displays "USB", dmesg indicates a device reset, the device gets back on the bus under the same product ID and can be accessed again. This usually happens on the first attempt after connecting the camera, unless it's a timeout (2.)
```
No native drivers available
Using drivers libusb-MSC, libusb-MTP
Looking for Sony devices

Querying mass storage device
Sony Camcorder is a camera in mass storage mode

Initializing firmware update
^CTraceback (most recent call last):
  File "pmca-console.py", line 67, in <module>
    main()
  File "pmca-console.py", line 61, in main
    firmwareUpdateCommand(args.datFile, args.driver)
  File "/home/enleth/sony/Sony-PMCA-RE/pmca/commands/usb.py", line 254, in firmwareUpdateCommand
    dev.checkGuard(file, size)
  File "/home/enleth/sony/Sony-PMCA-RE/pmca/usb/sony.py", line 320, in checkGuard
    self._sendWriteCommands(self.CMD_CHK_GUARD, file, size)
  File "/home/enleth/sony/Sony-PMCA-RE/pmca/usb/sony.py", line 291, in _sendWriteCommands
    windowSize, status = self._parseWriteResponse(self._sendCommand(command, writeParam + data))
  File "/home/enleth/sony/Sony-PMCA-RE/pmca/usb/sony.py", line 271, in _sendCommand
    response = self.dev.sendSonyExtCommand(self.SONY_CMD_Updater, commandHeader + data, bufferSize)
  File "/home/enleth/sony/Sony-PMCA-RE/pmca/usb/sony.py", line 69, in sendSonyExtCommand
    response, data = self.driver.sendReadCommand(command, bufferSize)
  File "/home/enleth/sony/Sony-PMCA-RE/pmca/usb/driver/libusb.py", line 162, in sendReadCommand
    self._writeCommand(self.DIRECTION_READ, command, size)
  File "/home/enleth/sony/Sony-PMCA-RE/pmca/usb/driver/libusb.py", line 122, in _writeCommand
    command = command.ljust(16, b'\0'),
  File "/home/enleth/sony/Sony-PMCA-RE/pmca/usb/driver/libusb.py", line 104, in write
    return self.dev.write(self.epOut, data)
  File "/home/enleth/sony/Sony-PMCA-RE/.env/lib/python2.7/site-packages/usb/core.py", line 940, in write
    intf, ep = self._ctx.setup_request(self, endpoint)
  File "/home/enleth/sony/Sony-PMCA-RE/.env/lib/python2.7/site-packages/usb/core.py", line 102, in wrapper
    return f(self, *args, **kwargs)
  File "/home/enleth/sony/Sony-PMCA-RE/.env/lib/python2.7/site-packages/usb/core.py", line 216, in setup_request
    self.managed_claim_interface(device, intf)
  File "/home/enleth/sony/Sony-PMCA-RE/.env/lib/python2.7/site-packages/usb/core.py", line 102, in wrapper
    return f(self, *args, **kwargs)
  File "/home/enleth/sony/Sony-PMCA-RE/.env/lib/python2.7/site-packages/usb/core.py", line 159, in managed_claim_interface
    self.managed_open()
  File "/home/enleth/sony/Sony-PMCA-RE/.env/lib/python2.7/site-packages/usb/core.py", line 104, in wrapper
    self.lock.release()
  File "/usr/lib64/python2.7/threading.py", line 187, in release
    def release(self):
KeyboardInterrupt
```
1a. I probably managed to ^C it once while it wasn't actually stuck, as the result was:
```
No native drivers available
Using drivers libusb-MSC, libusb-MTP
Looking for Sony devices

Querying mass storage device
Sony Camcorder is a camera in mass storage mode

Initializing firmware update
^CTraceback (most recent call last):
  File "pmca-console.py", line 67, in <module>
    main()
  File "pmca-console.py", line 61, in main
    firmwareUpdateCommand(args.datFile, args.driver)
  File "/home/enleth/sony/Sony-PMCA-RE/pmca/commands/usb.py", line 254, in firmwareUpdateCommand
    dev.checkGuard(file, size)
  File "/home/enleth/sony/Sony-PMCA-RE/pmca/usb/sony.py", line 320, in checkGuard
    self._sendWriteCommands(self.CMD_CHK_GUARD, file, size)
  File "/home/enleth/sony/Sony-PMCA-RE/pmca/usb/sony.py", line 291, in _sendWriteCommands
    windowSize, status = self._parseWriteResponse(self._sendCommand(command, writeParam + data))
  File "/home/enleth/sony/Sony-PMCA-RE/pmca/usb/sony.py", line 271, in _sendCommand
    response = self.dev.sendSonyExtCommand(self.SONY_CMD_Updater, commandHeader + data, bufferSize)
  File "/home/enleth/sony/Sony-PMCA-RE/pmca/usb/sony.py", line 69, in sendSonyExtCommand
    response, data = self.driver.sendReadCommand(command, bufferSize)
  File "/home/enleth/sony/Sony-PMCA-RE/pmca/usb/driver/libusb.py", line 173, in sendReadCommand
    sense = self._readResponse(failOnError)
  File "/home/enleth/sony/Sony-PMCA-RE/pmca/usb/driver/libusb.py", line 126, in _readResponse
    response = MscCommandStatusWrapper.unpack(self.read(MscCommandStatusWrapper.size))
  File "/home/enleth/sony/Sony-PMCA-RE/pmca/usb/driver/libusb.py", line 101, in read
    return self.dev.read(self.epIn, length).tostring()
  File "/home/enleth/sony/Sony-PMCA-RE/.env/lib/python2.7/site-packages/usb/core.py", line 992, in read
    elif ret != len(buff) * buff.itemsize:
KeyboardInterrupt
```
and on the following attempts, I started getting the below error, until I either reconnected the camera or waited a few minutes. It was still in "USB" mode all the time, did not switch to update mode.
```No native drivers available
Using drivers libusb-MSC, libusb-MTP
Looking for Sony devices

Querying mass storage device
Sony Camcorder is a camera in mass storage mode

Initializing firmware update
Traceback (most recent call last):
  File "pmca-console.py", line 67, in <module>
    main()
  File "pmca-console.py", line 61, in main
    firmwareUpdateCommand(args.datFile, args.driver)
  File "/home/enleth/sony/Sony-PMCA-RE/pmca/commands/usb.py", line 254, in firmwareUpdateCommand
    dev.checkGuard(file, size)
  File "/home/enleth/sony/Sony-PMCA-RE/pmca/usb/sony.py", line 320, in checkGuard
    self._sendWriteCommands(self.CMD_CHK_GUARD, file, size)
  File "/home/enleth/sony/Sony-PMCA-RE/pmca/usb/sony.py", line 291, in _sendWriteCommands
    windowSize, status = self._parseWriteResponse(self._sendCommand(command, writeParam + data))
  File "/home/enleth/sony/Sony-PMCA-RE/pmca/usb/sony.py", line 271, in _sendCommand
    response = self.dev.sendSonyExtCommand(self.SONY_CMD_Updater, commandHeader + data, bufferSize)
  File "/home/enleth/sony/Sony-PMCA-RE/pmca/usb/sony.py", line 62, in sendSonyExtCommand
    self._checkResponse(response)
  File "/home/enleth/sony/Sony-PMCA-RE/pmca/usb/__init__.py", line 20, in _checkResponse
    raise Exception('Mass storage error: Sense 0x%x 0x%x 0x%x' % sense)
Exception: Mass storage error: Sense 0x9 0x81 0x83
```
I'm not sure if this case is of any interest as this was most likely user error.

2. Starting out same way as above, but getting "Operation timed out", almost instantaneously
```No native drivers available
Using drivers libusb-MSC, libusb-MTP
Looking for Sony devices

Querying mass storage device
Traceback (most recent call last):
  File "pmca-console.py", line 67, in <module>
    main()
  File "pmca-console.py", line 61, in main
    firmwareUpdateCommand(args.datFile, args.driver)
  File "/home/enleth/sony/Sony-PMCA-RE/pmca/commands/usb.py", line 243, in firmwareUpdateCommand
    device = getDevice(driver)
  File "/home/enleth/sony/Sony-PMCA-RE/pmca/commands/usb.py", line 164, in getDevice
    devices = list(listDevices(driver))
  File "/home/enleth/sony/Sony-PMCA-RE/pmca/commands/usb.py", line 142, in listDevices
    info = MscDevice(drv).getDeviceInfo()
  File "/home/enleth/sony/Sony-PMCA-RE/pmca/usb/__init__.py", line 16, in __init__
    self.reset()
  File "/home/enleth/sony/Sony-PMCA-RE/pmca/usb/__init__.py", line 24, in reset
    self.driver.sendCommand(6 * b'\0')
  File "/home/enleth/sony/Sony-PMCA-RE/pmca/usb/driver/libusb.py", line 143, in sendCommand
    return self._readResponse(failOnError)
  File "/home/enleth/sony/Sony-PMCA-RE/pmca/usb/driver/libusb.py", line 126, in _readResponse
    response = MscCommandStatusWrapper.unpack(self.read(MscCommandStatusWrapper.size))
  File "/home/enleth/sony/Sony-PMCA-RE/pmca/usb/driver/libusb.py", line 101, in read
    return self.dev.read(self.epIn, length).tostring()
  File "/home/enleth/sony/Sony-PMCA-RE/.env/lib/python2.7/site-packages/usb/core.py", line 988, in read
    self.__get_timeout(timeout))
  File "/home/enleth/sony/Sony-PMCA-RE/.env/lib/python2.7/site-packages/usb/backend/libusb1.py", line 833, in bulk_read
    timeout)
  File "/home/enleth/sony/Sony-PMCA-RE/.env/lib/python2.7/site-packages/usb/backend/libusb1.py", line 936, in __read
    _check(retval)
  File "/home/enleth/sony/Sony-PMCA-RE/.env/lib/python2.7/site-packages/usb/backend/libusb1.py", line 595, in _check
    raise USBError(_strerror(ret), ret, _libusb_errno[ret])
usb.core.USBError: [Errno 110] Operation timed out
``` 
3. Camera connected and turned on, displays "USB", running `pmca-console firmware -f firmware_packed.dat`
```
No native drivers available
Using drivers libusb-MSC, libusb-MTP
Looking for Sony devices

Querying mass storage device
Sony Camcorder is a camera in mass storage mode

Initializing firmware update
Updating from version 1.10 to version 1.11
Switching to updater mode
Please press Ok to reset the camera, then run this command again to install the firmware
```
At this point the camera display switches to "UPDAT" for 2-3 seconds, then "USB" for a split second, then goes blank, the rear status LED goes red and stays this way. Obviously I'm not pressing the "Ok" button as there isn't any, it restarts by itself. dmesg reports a USB disconnect, then a connect under a different product ID. A single block device appears, but trying to access it returns -ENOMEDIUM. New USB device details are:
```Bus 002 Device 040: ID 054c:03e2 Sony Corp. 
Device Descriptor:
  bLength                18
  bDescriptorType         1
  bcdUSB               2.00
  bDeviceClass            0 
  bDeviceSubClass         0 
  bDeviceProtocol         0 
  bMaxPacketSize0        64
  idVendor           0x054c Sony Corp.
  idProduct          0x03e2 
  bcdDevice            1.00
  iManufacturer           1 Sony
  iProduct                2 Sony USB Device
  iSerial                 3 6C092035EFAB
  bNumConfigurations      1
  Configuration Descriptor:
    bLength                 9
    bDescriptorType         2
    wTotalLength           32
    bNumInterfaces          1
    bConfigurationValue     1
    iConfiguration          0 
    bmAttributes         0xc0
      Self Powered
    MaxPower              100mA
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        0
      bAlternateSetting       0
      bNumEndpoints           2
      bInterfaceClass         8 Mass Storage
      bInterfaceSubClass      6 SCSI
      bInterfaceProtocol     80 Bulk-Only
      iInterface              0 
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x81  EP 1 IN
        bmAttributes            2
          Transfer Type            Bulk
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0200  1x 512 bytes
        bInterval               0
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x02  EP 2 OUT
        bmAttributes            2
          Transfer Type            Bulk
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0200  1x 512 bytes
        bInterval               0
Device Qualifier (for other device speed):
  bLength                10
  bDescriptorType         6
  bcdUSB               2.00
  bDeviceClass            0 
  bDeviceSubClass         0 
  bDeviceProtocol         0 
  bMaxPacketSize0        64
  bNumConfigurations      1
can't get debug descriptor: Resource temporarily unavailable
Device Status:     0x0001
  Self Powered
```
I assume this is the firmware update mode and so far everything is like it should be. I run `pmca-console firmware -f firmware_packed.dat` again and so far I'm getting those:

3a. Getting "Entity not found" with a device disconnect/reconnect:
```No native drivers available
Using drivers libusb-MSC, libusb-MTP
Looking for Sony devices

Querying mass storage device
Traceback (most recent call last):
  File "pmca-console.py", line 67, in <module>
    main()
  File "pmca-console.py", line 61, in main
    firmwareUpdateCommand(args.datFile, args.driver)
  File "/home/enleth/sony/Sony-PMCA-RE/pmca/commands/usb.py", line 243, in firmwareUpdateCommand
    device = getDevice(driver)
  File "/home/enleth/sony/Sony-PMCA-RE/pmca/commands/usb.py", line 164, in getDevice
    devices = list(listDevices(driver))
  File "/home/enleth/sony/Sony-PMCA-RE/pmca/commands/usb.py", line 142, in listDevices
    info = MscDevice(drv).getDeviceInfo()
  File "/home/enleth/sony/Sony-PMCA-RE/pmca/usb/__init__.py", line 16, in __init__
    self.reset()
  File "/home/enleth/sony/Sony-PMCA-RE/pmca/usb/__init__.py", line 23, in reset
    self.driver.reset()
  File "/home/enleth/sony/Sony-PMCA-RE/pmca/usb/driver/libusb.py", line 98, in reset
    self.dev.reset()
  File "/home/enleth/sony/Sony-PMCA-RE/.env/lib/python2.7/site-packages/usb/core.py", line 915, in reset
    self._ctx.backend.reset_device(self._ctx.handle)
  File "/home/enleth/sony/Sony-PMCA-RE/.env/lib/python2.7/site-packages/usb/backend/libusb1.py", line 893, in reset_device
    _check(self.lib.libusb_reset_device(dev_handle.handle))
  File "/home/enleth/sony/Sony-PMCA-RE/.env/lib/python2.7/site-packages/usb/backend/libusb1.py", line 595, in _check
    raise USBError(_strerror(ret), ret, _libusb_errno[ret])
usb.core.USBError: [Errno 2] Entity not found
```
dmesg:
```[148926.445319] usb 2-1.2: reset high-speed USB device number 40 using xhci_hcd
[148926.580726] usb 2-1.2: can't restore configuration #1 (error=-71)
[148926.580874] usb 2-1.2: USB disconnect, device number 40
[148927.325064] usb 2-1.2: new high-speed USB device number 42 using xhci_hcd
[148927.456248] usb 2-1.2: New USB device found, idVendor=054c, idProduct=03e2
[148927.456250] usb 2-1.2: New USB device strings: Mfr=1, Product=2, SerialNumber=3
[148927.456251] usb 2-1.2: Product: Sony USB Device
[148927.456252] usb 2-1.2: Manufacturer: Sony
[148927.456253] usb 2-1.2: SerialNumber: 6C092035EFAB
[148927.457111] usb-storage 2-1.2:1.0: USB Mass Storage device detected
[148927.457364] scsi host3: usb-storage 2-1.2:1.0
[148928.575119] usb 2-1.2: reset high-speed USB device number 42 using xhci_hcd
[148929.705103] scsi 3:0:0:0: Direct-Access     Sony     Camcorder        1.00 PQ: 0 ANSI: 0
[148929.705365] sd 3:0:0:0: Attached scsi generic sg1 type 0
[148929.745269] sd 3:0:0:0: [sdb] Attached SCSI removable disk
```
3b. Getting "Entity not found" with a permanent device disconnect - the rear LED is still on and red, the screen is blank, I have to disconnect the USB cable and take the battery out to reboot the camera; then switch to update mode again. Most of the time I'm consistently getting 3a first, then 3b, but at least once I got 3a twice, then 3b, and at least once it was 3b immediately after switching to update mode.
```No native drivers available
Using drivers libusb-MSC, libusb-MTP
Looking for Sony devices

Querying mass storage device
Traceback (most recent call last):
  File "pmca-console.py", line 67, in <module>
    main()
  File "pmca-console.py", line 61, in main
    firmwareUpdateCommand(args.datFile, args.driver)
  File "/home/enleth/sony/Sony-PMCA-RE/pmca/commands/usb.py", line 243, in firmwareUpdateCommand
    device = getDevice(driver)
  File "/home/enleth/sony/Sony-PMCA-RE/pmca/commands/usb.py", line 164, in getDevice
    devices = list(listDevices(driver))
  File "/home/enleth/sony/Sony-PMCA-RE/pmca/commands/usb.py", line 142, in listDevices
    info = MscDevice(drv).getDeviceInfo()
  File "/home/enleth/sony/Sony-PMCA-RE/pmca/usb/__init__.py", line 16, in __init__
    self.reset()
  File "/home/enleth/sony/Sony-PMCA-RE/pmca/usb/__init__.py", line 23, in reset
    self.driver.reset()
  File "/home/enleth/sony/Sony-PMCA-RE/pmca/usb/driver/libusb.py", line 98, in reset
    self.dev.reset()
  File "/home/enleth/sony/Sony-PMCA-RE/.env/lib/python2.7/site-packages/usb/core.py", line 915, in reset
    self._ctx.backend.reset_device(self._ctx.handle)
  File "/home/enleth/sony/Sony-PMCA-RE/.env/lib/python2.7/site-packages/usb/backend/libusb1.py", line 893, in reset_device
    _check(self.lib.libusb_reset_device(dev_handle.handle))
  File "/home/enleth/sony/Sony-PMCA-RE/.env/lib/python2.7/site-packages/usb/backend/libusb1.py", line 595, in _check
    raise USBError(_strerror(ret), ret, _libusb_errno[ret])
usb.core.USBError: [Errno 2] Entity not found
```
dmesg:
```[149021.531387] usb 2-1.2: reset high-speed USB device number 42 using xhci_hcd
[149021.664602] usb 2-1.2: can't restore configuration #1 (error=-71)
[149021.664668] usb 2-1.2: USB disconnect, device number 42
```
3c. Getting "Input/output error" sometimes instead of "Entity not found", not sure if there's a pattern here.
```No native drivers available
Using drivers libusb-MSC, libusb-MTP
Looking for Sony devices

Querying mass storage device
Traceback (most recent call last):
  File "pmca-console.py", line 67, in <module>
    main()
  File "pmca-console.py", line 61, in main
    firmwareUpdateCommand(args.datFile, args.driver)
  File "/home/enleth/sony/Sony-PMCA-RE/pmca/commands/usb.py", line 243, in firmwareUpdateCommand
    device = getDevice(driver)
  File "/home/enleth/sony/Sony-PMCA-RE/pmca/commands/usb.py", line 164, in getDevice
    devices = list(listDevices(driver))
  File "/home/enleth/sony/Sony-PMCA-RE/pmca/commands/usb.py", line 142, in listDevices
    info = MscDevice(drv).getDeviceInfo()
  File "/home/enleth/sony/Sony-PMCA-RE/pmca/usb/__init__.py", line 16, in __init__
    self.reset()
  File "/home/enleth/sony/Sony-PMCA-RE/pmca/usb/__init__.py", line 24, in reset
    self.driver.sendCommand(6 * b'\0')
  File "/home/enleth/sony/Sony-PMCA-RE/pmca/usb/driver/libusb.py", line 143, in sendCommand
    return self._readResponse(failOnError)
  File "/home/enleth/sony/Sony-PMCA-RE/pmca/usb/driver/libusb.py", line 126, in _readResponse
    response = MscCommandStatusWrapper.unpack(self.read(MscCommandStatusWrapper.size))
  File "/home/enleth/sony/Sony-PMCA-RE/pmca/usb/driver/libusb.py", line 101, in read
    return self.dev.read(self.epIn, length).tostring()
  File "/home/enleth/sony/Sony-PMCA-RE/.env/lib/python2.7/site-packages/usb/core.py", line 988, in read
    self.__get_timeout(timeout))
  File "/home/enleth/sony/Sony-PMCA-RE/.env/lib/python2.7/site-packages/usb/backend/libusb1.py", line 833, in bulk_read
    timeout)
  File "/home/enleth/sony/Sony-PMCA-RE/.env/lib/python2.7/site-packages/usb/backend/libusb1.py", line 936, in __read
    _check(retval)
  File "/home/enleth/sony/Sony-PMCA-RE/.env/lib/python2.7/site-packages/usb/backend/libusb1.py", line 595, in _check
    raise USBError(_strerror(ret), ret, _libusb_errno[ret])
usb.core.USBError: [Errno 5] Input/Output Error
```
dmesg:
```[151609.194524] usb 2-1.2: reset high-speed USB device number 55 using xhci_hcd
[151609.545702] usb 2-1.2: USB disconnect, device number 55
[151610.094303] usb 2-1.2: new high-speed USB device number 56 using xhci_hcd
[151610.225518] usb 2-1.2: New USB device found, idVendor=054c, idProduct=03e2
[151610.225520] usb 2-1.2: New USB device strings: Mfr=1, Product=2, SerialNumber=3
[151610.225521] usb 2-1.2: Product: Sony USB Device
[151610.225522] usb 2-1.2: Manufacturer: Sony
[151610.225523] usb 2-1.2: SerialNumber: 6C092035EFAB
[151610.226333] usb-storage 2-1.2:1.0: USB Mass Storage device detected
[151610.226414] scsi host3: usb-storage 2-1.2:1.0
[151611.244854] scsi 3:0:0:0: Direct-Access     Sony     Camcorder        1.00 PQ: 0 ANSI: 0
[151611.244994] sd 3:0:0:0: Attached scsi generic sg1 type 0
[151611.246249] sd 3:0:0:0: [sdb] Attached SCSI removable disk
```
So far every time this one happened, the next attempt generated an "Entity not found" error and made the camera disconnect completely.

Also, I noticed one thing that could be important: rarely when 3a or 3b happens, there's also a "Sony Camcorder is a camera in mass storage mode" line in the output - that is, while the camera is in update mode. The traceback is identical, it doesn't get past pmca/commands/usb.py:243. I'm not sure what makes it print that line or omit it, but most of the time it's not there.

I never got to the firmware upload stage.

BTW, does the updaterbody library have any means of sending some text to the firmware update tool? This camera doesn't even have a proper display, there are no test pads on the PCBs that could be a serial port (I took it apart completely to check before I started playing with the updater) and what appears to be a complete kernel command line that I found somewhere in the unpacked firmware suggests that hardware console is explicitly turned off (console=ttynull, an interesting console type that, when googled, is only mentioned in a single code snippet with comments in Chinese).

Other ideas I had to get some kind of output was to a) assume the SD card is mounted rw during the update (or try some shooting in the dark to get it to be mounted) and output to a file there, or maybe b) try to get the WiFi AP up (by the means of shooting in the dark as well) and start a telnetd server, or c) start adbd and hope it's able to work in update mode - but I expect all of that to take a long time and be very painful with no feedback when something doesn't work. Any suggestions?

**Comment by ma1co** [comment]
We can tell it's a gen3 device from the firmware update (see _config.yaml_). Basically everthing released after august 2013 is gen3.

Oh god, that really doesn't sound very stable...

Can you try to add a short delay (something like `import time; time.sleep(.1)`) in the `read` and `write` functions in `_UsbDriver` in _pmca/usb/driver/libusb.py_, just to slow everything down?

Once the script crashes, you can't really run it again deterministically without deconnecting your camera first, you'll just get random errors.

You're correct, case 3 with idProduct=0x03e2 is the firmware update mode, so that's good.

The script then seems to crash always during the device reset. The reset is not strictly needed, so you could try to remove the call to `self.reset()` in the `__init__` function of `MscDevice` in _pmca/usb/\_\_init\_\_.py_.

Let's hope we get the updater to work...

Sadly, there is no easy way to send data back from the camera, not even the SD card drivers are loaded in update mode (for gen3 devices). Blinking the LED is about all we can do.

I would suggest to use the firmware updater to try to unlock app install mode (see issue #12, but I now have some ideas how to do that), then replace the SmartRemote app (see [this](https://github.com/ma1co/OpenMemories-Tweak/issues/58#issuecomment-272558715)) with something that starts telnetd. You could then telnet into the running system. Android is the most reliable way to run code atm...

But first, we need to get the USB drivers working!
